core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -553,12 +562,14 @@ private[spark] class ExecutorAllocationManager(
         }
 
         // If this is the last pending task, mark the scheduler queue as empty
-        stageIdToTaskIndices.getOrElseUpdate(stageId, new mutable.HashSet[Int]) += taskIndex
+        stageIdToTaskIndices
+          .getOrElseUpdate(stageId, new mutable.HashSet[String]) += (taskIndex + "." + attemptId)
         val numTasksScheduled = stageIdToTaskIndices(stageId).size
         val numTasksTotal = stageIdToNumTasks.getOrElse(stageId, -1)
         if (numTasksScheduled == numTasksTotal) {
           // No more pending tasks for this stage
           stageIdToNumTasks -= stageId
```

Comment:
sorry, I am wrong the line number, actually they are [these code](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala#L567-L574)